lib/rucio/common/types.py

```diff
@@ -99,3 +102,59 @@ class InternalScope(InternalType):
     '''
     def __init__(self, scope, vo='def', fromExternal=True):
         super(InternalScope, self).__init__(value=scope, vo=vo, fromExternal=fromExternal)
+
+
+class RSEDomainLANDict(TypedDict):
+    read: Optional[int]
+    write: Optional[int]
+    delete: Optional[int]
+
+
+class RSEDomainWANDict(TypedDict):
+    read: Optional[int]
+    write: Optional[int]
+    delete: Optional[int]
+    third_party_copy_read: Optional[int]
+    third_party_copy_write: Optional[int]
+
+
+class RSEDomainsDict(TypedDict):
+    lan: RSEDomainLANDict
+    wan: RSEDomainWANDict
+
+
+class RSEProtocolDict(TypedDict):
+    auth_token: Optional[str]  # FIXME: typing.NotRequired
+    hostname: Optional[str]
+    scheme: Optional[str]
+    port: Optional[int]
+    prefix: Optional[str]
+    impl: Optional[str]
+    domains: RSEDomainsDict
+    extended_attributes: Optional[Union[str, dict[str, Any]]]
```

Comment:
I think you’ve overused `Optional`. None of `hostname`, `scheme`, `port` or `impl` may be allowed to be `None`.

Reply:
In addition, could you point me to where `auth_token` is coming from?

Reply:
Sure, 
https://github.com/rucio/rucio/blob/7f0d229ac0b3bc7dec12c6e158bea2b82d414a3b/lib/rucio/rse/rsemanager.py#L189
this line of code made one of the github checks fail so I decided it is a good idea to include `auth_token`.

Reply:
Thank you. This is one of those unfortunate cases where the dictionary structure changes as it’s passed between layers. I think this is a prime case for `typing.NotRequired`, as you had before.